modules/page_object_about_newtab.py

```diff
@@ -0,0 +1,130 @@
+        WebDriver object under test
+
+    lang_code: str
+        The ISO-639-1 code of the language, followed by the ISO-3166-1 alpha-2 code
+        of the region (country). examples: enUS, zhCN, jaJP, frFR
+
+    constants: self.L10nAboutNewtab
+        Instantiation of the constants object
+    """
+
+    URL_TEMPLATE = "about:newtab"
+
+    def set_language_code(self, lang_code: str) -> BasePage:
+        """
+        Set the language code of the object, set self.constants as well
+        """
+        self.language = lang_code
+        self.constants = L10nAboutNewtab(lang_code)
+        return self
+
+    def count_top_sites(self) -> int:
+        """
+        Find the top sites area and return the number of tiles in it
+        """
+        top_sites_list = self.get_element("top-sites-list")
+        top_site_nodes = top_sites_list.find_elements(By.TAG_NAME, "li")
+        return len(top_site_nodes)
+
+    def get_recommended_articles(self) -> list[WebElement]:
+        """
+        Return a list of all recommended article nodes
+        """
+        article_root = self.get_element("recommended-by-pocket-list")
+        return article_root.find_elements(By.TAG_NAME, "article")
+
+    def check_article_alignment(self) -> bool:
+        """
+        Check that all article card elements are lined up with the elements in their columns
+        """
+        recd_articles = self.get_recommended_articles()
+        number_of_articles = len(recd_articles)
+        for i in range(number_of_articles - 3, number_of_articles):
+            article_x = recd_articles[i].location.get("x")
+            for j in range(i - 3, -1, -3):
+                comparison_x = recd_articles[j].location.get("x")
+                assert article_x == comparison_x
+        return True
+
+    def count_sponsored_articles(self) -> int:
+        """
+        Return the total number of sponsored article cards
+        """
+        return len(self.get_elements("story-sponsored-footer"))
+
+    def check_popular_topics(self) -> bool:
+        """
+        Ensure that all the Popular Topics are as expected
+        """
+        poptopics_list = self.get_element("popular-topics-list")
+        checked_topics = []
+        logging.info(f"Topics: {self.constants.popular_topics}")
+        for li in poptopics_list.find_elements(By.TAG_NAME, "li"):
+            this_topic = li.get_attribute("innerText")
+            logging.info(f"Topic innerText = {this_topic}")
+            assert this_topic in self.constants.popular_topics
+            checked_topics.append(this_topic)
+        assert checked_topics == self.constants.popular_topics
+        return True
+
+    def check_layout(self) -> BasePage:
+        """
+        Main method to check about:newtab layout is as expected
+        """
+        self.element_exists("incontent-search-input")
+        logging.info("Search bar exists")
+        self.element_exists("recent-activity-section")
+
+        for i in range(1, 22):
```

Comment:
why specifically 22 times? it would be nice to define this as a constant at the top for less _magical numberesque_ ness if you get me

Reply:
Will demystify this.